rfcs/20200612-stream-executor-c-api.md

```diff
@@ -0,0 +1,557 @@
+
+```cpp
+static bool IsMyCustomPlatformRegistered = []() {
+ RegisterMyCustomPlatform();
+ return true;
+}();
+```
+
+## Stream/Timer/Event representation
+
+API extension would require defining SE\_Stream\_st, SE\_Event\_st and
+SE\_Timer\_st structs. From the point of view of TensorFlow, we will treat their
+pointers as opaque.
+
+Underneath, StreamExecutor will rely on customized implementations of
+[StreamInterface](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/stream_executor/stream_executor_internal.h;l=114?q=TimerInterface&ss=tensorflow%2Ftensorflow),
+[TimerInterface](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/stream_executor/stream_executor_internal.h;l=145?q=TimerInterface&ss=tensorflow%2Ftensorflow)
+and
+[EventInterface](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/stream_executor/stream_executor_internal.h;l=76?q=TimerInterface&ss=tensorflow%2Ftensorflow).
+For example, Stream customization might look as follows:
+
+```cpp
+class CStream : public StreamInterface {
+ public:
+  explicit CStream(SE_Device* device,
+                   SE_StreamExecutor* stream_executor) :
+    device_(device), stream_executor_(stream_executor),
+    stream_handle_(nullptr) {
+  }
+  ~CStream() override {
+    Destroy();
+  }
+
+  void Init() {
+    stream_handle_ = stream_executor_->create_stream(device_);
+  }
+
+  void Destroy() {
+    if (stream_handle_ != nullptr) {
+      stream_executor_->delete_stream(device_, stream_handle_);
+      stream_handle_ = nullptr;
+    }
+  }
+
+  SE_Stream Handle() {
+    return stream_handle_;
+  }
+
+ private:
+  SE_Device* device_;  // not owned
+  SE_StreamExecutor* stream_executor_;  // not owned
+  SE_Stream stream_handle_;
+};
+```
+
+## Stability / User Impact
+
+The C API will be placed under _tensorflow/c/experimental/_ directory.
+Initially, we won’t have any compatibility guarantees. At the same time we will
+make the best effort to perform any updates in a backwards compatible way. For
+e.g. we plan to keep track of struct sizes.
+
+We will have an initial bake-in period before we consider moving the API out of experimental directory.
+
+## Alternatives Considered
+
+*   **Forking:** Contributors could always fork the TensorFlow repository,
+    directly make changes there to add a device, and release custom TensorFlow
+    packages. However, keeping forked copy in sync with the main repository can
+    be challenging and tedious, especially if some breakages cannot be fixed and
+    the code diverges.
+*   **Designing a new C API instead of StreamExecutor:** We are transitioning to
+    the new TensorFlow stack soon. Since the current stack’s code might not be
+    compatible with the new stack, we decided to stick with the existing
+    StreamExecutorInterface to minimize throw-away efforts.
+
+## Performance Implications
+
+The C API should not affect TensorFlow’s performance. Using the C API to connect
```

Comment:
It can theoretically have an effect right, since we now have another level of indirection (going through a function pointer)?

Reply:
We would only go through function pointer for the plugins though. These plugins don't exist now, so this shouldn't impact performance of existing stack.

Reply:
Based on design review, seems like performance is not a concern here.